main.go

```diff
@@ -160,10 +162,33 @@ func main() {
 	server := &http.Server{
 		Addr: cfg.listenAddr,
 	}
-	if err := serve(logger, cfg, server, writers, readers); err != nil {
-		_ = level.Error(logger).Log("msg", "Failed to listen", "addr", cfg.listenAddr, "err", err)
+
+	done := make(chan os.Signal, 1)
+	signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		if err := serve(logger, cfg, server, writers, readers); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			_ = level.Error(logger).Log("msg", "Failed to listen", "addr", cfg.listenAddr, "err", err)
+			os.Exit(1)
+		}
+	}()
+
+	<-done
+
+	// first asking http to shut down and wait for it up to 10 seconds
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	errShutdown := server.Shutdown(ctx)
+
+	// after shutdown of server (successfull or not) close clients
+	closeClients(logger, writers, readers)
+
+	if err := errShutdown; err != nil {
```

Comment:
Why do you assign errShutdown to a new variable?

Reply:
I guess because I copy and pasted :slightly_smiling_face: